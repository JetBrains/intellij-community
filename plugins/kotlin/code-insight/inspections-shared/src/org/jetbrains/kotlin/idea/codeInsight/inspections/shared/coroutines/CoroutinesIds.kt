// Copyright 2000-2025 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package org.jetbrains.kotlin.idea.codeInsight.inspections.shared.coroutines

import org.jetbrains.kotlin.name.CallableId
import org.jetbrains.kotlin.name.ClassId
import org.jetbrains.kotlin.name.FqName
import org.jetbrains.kotlin.name.Name

internal object CoroutinesIds {
    val COROUTINES_PACKAGE: FqName = FqName("kotlinx.coroutines")

    val CURRENT_COROUTINE_CONTEXT_ID: CallableId = CallableId(COROUTINES_PACKAGE, Name.identifier("currentCoroutineContext"))
    
    val JOB_CLASS_ID: ClassId = ClassId(COROUTINES_PACKAGE, Name.identifier("Job"))
    val JOB_JOIN_ID: CallableId = CallableId(JOB_CLASS_ID, Name.identifier("join"))
    val JOIN_ALL_ID: CallableId = CallableId(COROUTINES_PACKAGE, Name.identifier("joinAll"))
    
    val DEFERRED_CLASS_ID: ClassId = ClassId(COROUTINES_PACKAGE, Name.identifier("Deferred"))
    val DEFERRED_AWAIT_ID: CallableId = CallableId(DEFERRED_CLASS_ID, Name.identifier("await"))
    val AWAIT_ALL_ID: CallableId = CallableId(COROUTINES_PACKAGE, Name.identifier("awaitAll"))
    
    val COROUTINE_SCOPE_CLASS_ID: ClassId = ClassId(COROUTINES_PACKAGE, Name.identifier("CoroutineScope"))
    
    val COROUTINES_SELECTS_PACKAGE: FqName = FqName("kotlinx.coroutines.selects")

    val SELECT_BUILDER_CLASS_ID: ClassId = ClassId(COROUTINES_SELECTS_PACKAGE, Name.identifier("SelectBuilder"))
    val SELECT_BUILDER_INVOKE_ID: CallableId = CallableId(SELECT_BUILDER_CLASS_ID, Name.identifier("invoke"))
    val SELECT_BUILDER_ON_TIMEOUT_ID: CallableId = CallableId(COROUTINES_SELECTS_PACKAGE, Name.identifier("onTimeout"))
}
