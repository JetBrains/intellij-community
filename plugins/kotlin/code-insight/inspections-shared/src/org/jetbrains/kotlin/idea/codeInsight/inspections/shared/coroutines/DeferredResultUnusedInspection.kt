// Copyright 2000-2025 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package org.jetbrains.kotlin.idea.codeInsight.inspections.shared.coroutines

import com.intellij.codeInspection.LocalInspectionToolSession
import com.intellij.codeInspection.ProblemsHolder
import com.intellij.codeInspection.options.OptPane
import com.intellij.codeInspection.options.OptPane.checkbox
import com.intellij.codeInspection.options.OptPane.pane
import com.intellij.psi.PsiElement
import org.jetbrains.kotlin.analysis.api.KaSession
import org.jetbrains.kotlin.analysis.api.analyze
import org.jetbrains.kotlin.analysis.api.resolution.KaFunctionCall
import org.jetbrains.kotlin.analysis.api.resolution.singleFunctionCallOrNull
import org.jetbrains.kotlin.analysis.api.resolution.symbol
import org.jetbrains.kotlin.idea.base.resources.KotlinBundle
import org.jetbrains.kotlin.idea.codeinsight.api.classic.inspections.AbstractKotlinInspection
import org.jetbrains.kotlin.name.FqName
import org.jetbrains.kotlin.psi.*

internal class DeferredResultUnusedInspection(@JvmField var standardOnly: Boolean = false) : AbstractKotlinInspection() {
    private fun isExpressionApplicable(expression: KtExpression): Boolean =
        expression is KtCallExpression && (!standardOnly || expression.calleeExpression?.text in shortNames)

    private fun KaSession.shouldReportCall(call: KaFunctionCall<*>): Boolean {
        val callableId = call.partiallyAppliedSymbol.symbol.callableId?.asSingleFqName()
        if (callableId in fqNamesThatShouldNotBeReported) return false

        return if (standardOnly) {
            callableId in fqNamesAll
        } else {
            val returnTypeClassId = call.partiallyAppliedSymbol.signature.returnType.expandedSymbol?.classId?.asSingleFqName()
            returnTypeClassId == deferred || returnTypeClassId == deferredExperimental
        }
    }

    private fun check(expression: KtExpression): Boolean {
        // Check whatever possible by PSI
        if (!isExpressionApplicable(expression)) return false
        var current: PsiElement? = expression
        var parent: PsiElement? = expression.parent
        while (parent != null) {
            if (parent is KtBlockExpression || parent is KtFunction || parent is KtFile) break
            if (parent is KtValueArgument || parent is KtBinaryExpression || parent is KtUnaryExpression) return false
            if (parent is KtQualifiedExpression && parent.receiverExpression == current) return false
            // TODO: add when condition, if condition (later when it's applicable not only to Deferred)
            current = parent
            parent = parent.parent
        }
        // Then check by call using Analysis API
        analyze(expression) {
            if (expression.isUsedAsExpression) return false
            val call = expression.resolveToCall()?.singleFunctionCallOrNull() ?: return false
            return shouldReportCall(call)
        }
    }

    override fun buildVisitor(
        holder: ProblemsHolder,
        isOnTheFly: Boolean,
        session: LocalInspectionToolSession
    ): KtVisitorVoid = callExpressionVisitor(fun(expression) {
        if (!check(expression)) return
        holder.registerProblem(expression.calleeExpression ?: expression, KotlinBundle.message("deferred.result.is.never.used"))
    })

    override fun getOptionsPane(): OptPane = pane(
        checkbox("standardOnly", KotlinBundle.message("reports.only.function.calls.from.kotlinx.coroutines")))
}

private const val COROUTINE_PACKAGE = "kotlinx.coroutines"

private const val COROUTINE_EXPERIMENTAL_PACKAGE = "kotlinx.coroutines.experimental"

private val shortNames = setOf("async")

private val fqNames: Set<FqName> = shortNames.mapTo(mutableSetOf()) { FqName("$COROUTINE_PACKAGE.$it") }

private val fqNamesExperimental: Set<FqName> = shortNames.mapTo(mutableSetOf()) { FqName("$COROUTINE_EXPERIMENTAL_PACKAGE.$it") }

private val fqNamesAll = fqNames + fqNamesExperimental

private val deferred: FqName = FqName("$COROUTINE_PACKAGE.Deferred")

private val deferredExperimental: FqName = FqName("$COROUTINE_EXPERIMENTAL_PACKAGE.Deferred")

private val fqNamesThatShouldNotBeReported =
    listOf("kotlin.test.assertNotNull", "kotlin.requireNotNull", "kotlin.checkNotNull").map { FqName(it) }
