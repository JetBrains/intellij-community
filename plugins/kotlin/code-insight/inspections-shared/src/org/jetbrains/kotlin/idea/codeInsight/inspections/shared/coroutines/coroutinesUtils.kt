// Copyright 2000-2025 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package org.jetbrains.kotlin.idea.codeInsight.inspections.shared.coroutines

import org.jetbrains.kotlin.analysis.api.KaSession
import org.jetbrains.kotlin.analysis.api.resolution.KaExplicitReceiverValue
import org.jetbrains.kotlin.analysis.api.resolution.successfulFunctionCallOrNull
import org.jetbrains.kotlin.analysis.api.resolution.successfulVariableAccessCall
import org.jetbrains.kotlin.analysis.api.resolution.symbol
import org.jetbrains.kotlin.analysis.api.symbols.KaFunctionSymbol
import org.jetbrains.kotlin.idea.codeinsight.utils.ConvertLambdaToReferenceUtils.singleStatementOrNull
import org.jetbrains.kotlin.name.CallableId
import org.jetbrains.kotlin.name.FqName
import org.jetbrains.kotlin.name.Name
import org.jetbrains.kotlin.name.StandardClassIds
import org.jetbrains.kotlin.psi.*
import org.jetbrains.kotlin.resolve.ImportPath

/**
 * Finds a single (implicitly or explicitly) returned expression from [lambdaExpression].
 */
internal fun KaSession.singleReturnedExpressionOrNull(lambdaExpression: KtLambdaExpression): KtExpression? {
    val singleStatement = lambdaExpression.singleStatementOrNull() ?: return null

    return when (singleStatement) {
        is KtReturnExpression if (singleStatement.targetSymbol == lambdaExpression.functionLiteral.symbol) -> singleStatement.returnedExpression
        else -> singleStatement
    }
}

/**
 * Checks if a lambda has a single parameter and calls a specific method on it.
 *
 * @param lambdaExpression The lambda expression to check
 * @param callableId The method that should be called on the parameter
 * @return true if the lambda has a single parameter and calls the specified method on it
 */
internal fun KaSession.isLambdaWithSingleReturnedCallOnSingleParameter(
    lambdaExpression: KtLambdaExpression,
    callableId: CallableId
): Boolean {
    val singleLambdaParameterSymbol = lambdaExpression.functionLiteral.symbol.valueParameters.singleOrNull() ?: return false
    val singleReturnedExpression = singleReturnedExpressionOrNull(lambdaExpression) as? KtDotQualifiedExpression ?: return false

    val methodCall = singleReturnedExpression.resolveToCall()?.successfulFunctionCallOrNull() ?: return false

    val explicitReceiverValue = methodCall.partiallyAppliedSymbol.dispatchReceiver as? KaExplicitReceiverValue ?: return false
    val explicitReceiverAccessCall = explicitReceiverValue.expression.resolveToCall()?.successfulVariableAccessCall() ?: return false

    return methodCall.symbol.callableId == callableId &&
            explicitReceiverAccessCall.symbol == singleLambdaParameterSymbol
}

/**
 * Checks if a package is already imported via star import.
 */
// TODO Use Import insertion API after KTIJ-28838 is fixed
internal fun isPackageImportedByStarImport(file: KtFile, packageFqName: FqName): Boolean {
    return file.importDirectives.any {
        it.importPath == ImportPath(packageFqName, isAllUnder = true)
    }
}

internal fun KaSession.isIterableForEachFunctionCall(element: KtCallExpression): Boolean {
    val functionCall = element.resolveToCall()?.successfulFunctionCallOrNull() ?: return false
    val actualReceiverType = functionCall.partiallyAppliedSymbol.extensionReceiver?.type ?: return false

    return isIterableForEachFunction(functionCall.symbol) &&
            actualReceiverType.isSubtypeOf(StandardClassIds.Collection)
}

private fun KaSession.isIterableForEachFunction(symbol: KaFunctionSymbol): Boolean {
    return symbol.callableId == KOTLIN_COLLECTIONS_FOR_EACH_ID &&
            symbol.receiverParameter?.returnType?.isSubtypeOf(StandardClassIds.Iterable) == true
}

internal fun KaSession.isIterableMapFunctionCall(element: KtCallExpression): Boolean {
    val functionCall = element.resolveToCall()?.successfulFunctionCallOrNull() ?: return false
    val actualReceiverType = functionCall.partiallyAppliedSymbol.extensionReceiver?.type ?: return false

    return isIterableMapFunction(functionCall.symbol) &&
            actualReceiverType.isSubtypeOf(StandardClassIds.Collection)
}

private fun KaSession.isIterableMapFunction(symbol: KaFunctionSymbol): Boolean {
    return symbol.callableId == KOTLIN_COLLECTIONS_MAP_ID &&
            symbol.receiverParameter?.returnType?.isSubtypeOf(StandardClassIds.Iterable) == true
}

// Common collection-related CallableIds
internal val KOTLIN_COLLECTIONS_FOR_EACH_ID: CallableId = CallableId(FqName("kotlin.collections"), Name.identifier("forEach"))
internal val KOTLIN_COLLECTIONS_MAP_ID: CallableId = CallableId(FqName("kotlin.collections"), Name.identifier("map"))
