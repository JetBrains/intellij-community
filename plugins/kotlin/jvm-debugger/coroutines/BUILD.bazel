### auto-generated section `build kotlin.jvm-debugger.coroutines` start
load("//build:compiler-options.bzl", "create_kotlinc_options")
load("@rules_jvm//:jvm.bzl", "jvm_library", "jvm_resources")

create_kotlinc_options(
  name = "custom",
  context_receivers = True,
  opt_in = ["org.jetbrains.kotlin.utils.addToStdlib.UnsafeCastFunction"]
)

jvm_resources(
  name = "kotlin-jvm-debugger-coroutines_resources",
  files = glob(["resources/**/*"]),
  strip_prefix = "resources"
)

jvm_library(
  name = "kotlin-jvm-debugger-coroutines",
  module_name = "kotlin.jvm-debugger.coroutines",
  visibility = ["//visibility:public"],
  srcs = glob(["src/**/*.kt", "src/**/*.java"], allow_empty = True),
  kotlinc_opts = ":custom",
  deps = [
    "@lib//:kotlin-stdlib",
    "@lib//:kotlinc-kotlin-compiler-common",
    "//java/debugger/openapi:java-debugger",
    "@lib//:kotlinc-analysis-api",
    "@lib//:maven-resolver-provider",
    "@lib//:gson",
    "@lib//:asm",
    "//plugins/kotlin/base/util:kotlin-base-util",
    "//plugins/kotlin/base/psi:kotlin-base-psi",
    "//plugins/kotlin/base/analysis:kotlin-base-analysis",
    "//plugins/kotlin/frontend-independent:fir-frontend-independent",
    "//plugins/kotlin/jvm-debugger/base/util:kotlin-jvm-debugger-base-util",
    "//plugins/kotlin/jvm-debugger/core:kotlin-jvm-debugger-core",
    "//platform/lang-impl",
    "//platform/xdebugger-impl:debugger-impl",
    "//platform/execution-impl",
    "//platform/external-system-impl:externalSystem-impl",
    "//platform/core-ui",
    "//platform/platform-util-io:ide-util-io",
    "//platform/smRunner",
    "//platform/threadDumpParser",
    "//java/java-psi-api:psi",
    "//platform/platform-impl:ide-impl",
    "//java/execution/impl:java-execution-impl",
    "//java/debugger/impl:java-debugger-impl",
    "//plugins/kotlin/code-insight/utils:kotlin-code-insight-utils",
    "//java/java-runtime:rt",
    "//plugins/kotlin/base/frontend-agnostic:kotlin-base-frontend-agnostic",
    "@lib//:kotlinx-serialization-core",
    "@lib//:kotlinx-serialization-json",
    "//java/java-impl:impl",
    "//java/debugger/shared:java-debugger-impl-shared",
  ],
  runtime_deps = [":kotlin-jvm-debugger-coroutines_resources"]
)
### auto-generated section `build kotlin.jvm-debugger.coroutines` end