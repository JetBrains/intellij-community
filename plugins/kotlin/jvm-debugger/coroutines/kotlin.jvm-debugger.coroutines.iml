<?xml version="1.0" encoding="UTF-8"?>
<module type="JAVA_MODULE" version="4">
  <component name="NewModuleRootManager" inherit-compiler-output="true">
    <exclude-output />
    <content url="file://$MODULE_DIR$">
      <sourceFolder url="file://$MODULE_DIR$/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/resources" type="java-resource" />
    </content>
    <orderEntry type="inheritedJdk" />
    <orderEntry type="sourceFolder" forTests="false" />
    <orderEntry type="library" name="kotlinc.kotlin-stdlib" level="project" />
    <orderEntry type="library" name="kotlinc.kotlin-compiler-common" level="project" />
    <orderEntry type="library" name="kotlinc.kotlin-compiler-fe10" level="project" />
    <orderEntry type="library" name="kotlinc.kotlin-compiler-ir" level="project" />
    <orderEntry type="library" name="kotlinc.high-level-api" level="project" />
    <orderEntry type="library" name="maven-resolver-provider" level="project" />
    <orderEntry type="library" name="gson" level="project" />
    <orderEntry type="library" name="ASM" level="project" />
    <orderEntry type="module" module-name="kotlin.base.util" />
    <orderEntry type="module" module-name="kotlin.base.psi" />
    <orderEntry type="module" module-name="kotlin.base.analysis" />
    <orderEntry type="module" module-name="kotlin.fir.frontend-independent" />
    <orderEntry type="module" module-name="kotlin.jvm-debugger.base.util" />
    <orderEntry type="module" module-name="kotlin.jvm-debugger.core" />
    <orderEntry type="module" module-name="intellij.platform.lang.impl" />
    <orderEntry type="module" module-name="intellij.platform.debugger.impl" />
    <orderEntry type="module" module-name="intellij.platform.execution.impl" />
    <orderEntry type="module" module-name="intellij.platform.externalSystem.impl" />
    <orderEntry type="module" module-name="intellij.platform.core.ui" />
    <orderEntry type="module" module-name="intellij.platform.ide.util.io" />
    <orderEntry type="module" module-name="intellij.platform.smRunner" />
    <orderEntry type="module" module-name="intellij.java.psi" />
    <orderEntry type="module" module-name="intellij.java.execution.impl" />
    <orderEntry type="module" module-name="intellij.java.debugger.impl" />
  </component>
</module>