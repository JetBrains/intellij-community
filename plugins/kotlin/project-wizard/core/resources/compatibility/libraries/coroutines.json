[
    {
        "ideVersion": "2023.1 EAP-INF",
        "groupId": "org.jetbrains.kotlinx",
        "artifactId": "kotlinx-coroutines-core",
        "versions": {
            "1.9": "1.8.0",
            "1.8": "1.7.3",
            "1.6": "1.6.4",
            "1.5": "1.5.2"
        }
    }
]