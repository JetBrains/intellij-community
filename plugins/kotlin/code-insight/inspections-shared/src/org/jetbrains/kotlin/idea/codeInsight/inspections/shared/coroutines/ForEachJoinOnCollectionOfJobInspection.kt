// Copyright 2000-2025 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package org.jetbrains.kotlin.idea.codeInsight.inspections.shared.coroutines

import com.intellij.codeInspection.ProblemsHolder
import com.intellij.codeInspection.util.InspectionMessage
import com.intellij.codeInspection.util.IntentionFamilyName
import com.intellij.modcommand.ModPsiUpdater
import com.intellij.openapi.project.Project
import org.jetbrains.kotlin.analysis.api.KaSession
import org.jetbrains.kotlin.idea.base.psi.imports.addImport
import org.jetbrains.kotlin.idea.base.resources.KotlinBundle
import org.jetbrains.kotlin.idea.codeInsight.inspections.shared.coroutines.CoroutinesIds.JOB_JOIN_ID
import org.jetbrains.kotlin.idea.codeInsight.inspections.shared.coroutines.CoroutinesIds.JOIN_ALL_ID
import org.jetbrains.kotlin.idea.codeinsight.api.applicable.inspections.KotlinApplicableInspectionBase
import org.jetbrains.kotlin.idea.codeinsight.api.applicable.inspections.KotlinModCommandQuickFix
import org.jetbrains.kotlin.idea.refactoring.singleLambdaArgumentExpression
import org.jetbrains.kotlin.psi.KtCallExpression
import org.jetbrains.kotlin.psi.KtPsiFactory
import org.jetbrains.kotlin.psi.KtVisitor
import org.jetbrains.kotlin.psi.callExpressionVisitor

/**
 * @see MapAwaitOnCollectionOfDeferredInspection for a similar inspection for `map { it.await() }` calls.
 */
internal class ForEachJoinOnCollectionOfJobInspection : KotlinApplicableInspectionBase.Simple<KtCallExpression, Unit>() {
    override fun getProblemDescription(
        element: KtCallExpression,
        context: Unit
    ): @InspectionMessage String {
        return KotlinBundle.message("inspection.foreach.join.on.collection.of.job.description")
    }

    override fun buildVisitor(
        holder: ProblemsHolder,
        isOnTheFly: Boolean
    ): KtVisitor<*, *> = callExpressionVisitor { callExpression ->
        visitTargetElement(callExpression, holder, isOnTheFly)
    }

    override fun KaSession.prepareContext(element: KtCallExpression): Unit? {
        val lambdaArgument = element.singleLambdaArgumentExpression() ?: return null

        if (!isIterableForEachFunctionCall(element)) return null
        if (!isLambdaWithSingleReturnedCallOnSingleParameter(lambdaArgument, JOB_JOIN_ID)) return null

        return Unit
    }

    override fun createQuickFix(
        element: KtCallExpression,
        context: Unit
    ): KotlinModCommandQuickFix<KtCallExpression> {
        return object : KotlinModCommandQuickFix<KtCallExpression>() {
            override fun getFamilyName(): @IntentionFamilyName String {
                return KotlinBundle.message("inspection.foreach.join.on.collection.of.job.replace.with.joinAll")
            }

            override fun applyFix(
                project: Project,
                element: KtCallExpression,
                updater: ModPsiUpdater
            ) {
                val alreadyImportedByStarImport = isPackageImportedByStarImport(element.containingKtFile, JOIN_ALL_ID.packageName)

                if (!alreadyImportedByStarImport) {
                    element.containingKtFile.addImport(JOIN_ALL_ID.asSingleFqName())
                }

                element.replace(KtPsiFactory(project).createExpression("${JOIN_ALL_ID.callableName}()"))
            }
        }
    }
}
