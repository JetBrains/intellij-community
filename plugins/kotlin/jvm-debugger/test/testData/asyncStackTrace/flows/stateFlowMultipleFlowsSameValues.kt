@file:JvmName("StateFlowMultipleFlowsSameValues.kt")
package flows.stateFlow.stateFlowMultipleFlowsSameValues.kt

// ATTACH_LIBRARY: maven(com.intellij.platform:kotlinx-coroutines-core-jvm:1.8.0-intellij-11)

import kotlinx.coroutines.channels.BufferOverflow
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withContext

val stateFlow1 = `~~~!!! recognizableFrameWithEmitInAsyncStackTraceConstructor 1 !!!~~~`()
val stateFlow2 = `~~~!!! recognizableFrameWithEmitInAsyncStackTraceConstructor 2 !!!~~~`()

fun main(): Unit = runBlocking {
    val job1 = launch {
        stateFlow1.collect {
            //Breakpoint!
            println(it)
        }
    }
    val job2 = launch {
        stateFlow2.collect {
            //Breakpoint!
            println(it)
        }
    }

    launch {
        `~~~!!! recognizableFrameWithEmitInAsyncStackTraceAssign 1 !!!~~~`(stateFlow1, 42)
        `~~~!!! recognizableFrameWithEmitInAsyncStackTraceAssign 2 !!!~~~`(stateFlow2, 42)
        delay(100)
        `~~~!!! recognizableFrameWithEmitInAsyncStackTraceEmit 1 !!!~~~`(stateFlow1, 81)
        `~~~!!! recognizableFrameWithEmitInAsyncStackTraceEmit 2 !!!~~~`(stateFlow2, 81)
        delay(100)
        `~~~!!! recognizableFrameWithEmitInAsyncStackTraceCas 1 !!!~~~`(stateFlow1, 81, 42)
        `~~~!!! recognizableFrameWithEmitInAsyncStackTraceCas 2 !!!~~~`(stateFlow2, 81, 42)
        delay(100)
        `~~~!!! recognizableFrameWithEmitInAsyncStackTraceAssign 3 !!!~~~`(stateFlow1, 81)
        `~~~!!! recognizableFrameWithEmitInAsyncStackTraceAssign 4 !!!~~~`(stateFlow2, 81)
        delay(100)
        `~~~!!! recognizableFrameWithEmitInAsyncStackTraceEmit 3 !!!~~~`(stateFlow1, 42)
        `~~~!!! recognizableFrameWithEmitInAsyncStackTraceEmit 4 !!!~~~`(stateFlow2, 42)
        delay(100)
        job1.cancel()
        job2.cancel()
    }
}

fun `~~~!!! recognizableFrameWithEmitInAsyncStackTraceConstructor 1 !!!~~~`(): MutableStateFlow<Int> {
    return MutableStateFlow(-1)
}

fun `~~~!!! recognizableFrameWithEmitInAsyncStackTraceConstructor 2 !!!~~~`(): MutableStateFlow<Int> {
    return MutableStateFlow(-1)
}


suspend fun <T> `~~~!!! recognizableFrameWithEmitInAsyncStackTraceAssign 1 !!!~~~`(flow: MutableStateFlow<T>, value: T) {
    flow.value = value
}

suspend fun <T> `~~~!!! recognizableFrameWithEmitInAsyncStackTraceEmit 1 !!!~~~`(flow: MutableStateFlow<T>, value: T) {
    flow.emit(value)
}

suspend fun <T> `~~~!!! recognizableFrameWithEmitInAsyncStackTraceCas 1 !!!~~~`(flow: MutableStateFlow<T>, old: T, new: T) {
    flow.compareAndSet(old, new)
}

suspend fun <T> `~~~!!! recognizableFrameWithEmitInAsyncStackTraceAssign 2 !!!~~~`(flow: MutableStateFlow<T>, value: T) {
    flow.value = value
}

suspend fun <T> `~~~!!! recognizableFrameWithEmitInAsyncStackTraceEmit 2 !!!~~~`(flow: MutableStateFlow<T>, value: T) {
    flow.emit(value)
}

suspend fun <T> `~~~!!! recognizableFrameWithEmitInAsyncStackTraceCas 2 !!!~~~`(flow: MutableStateFlow<T>, old: T, new: T) {
    flow.compareAndSet(old, new)
}

suspend fun <T> `~~~!!! recognizableFrameWithEmitInAsyncStackTraceAssign 3 !!!~~~`(flow: MutableStateFlow<T>, value: T) {
    flow.value = value
}

suspend fun <T> `~~~!!! recognizableFrameWithEmitInAsyncStackTraceEmit 3 !!!~~~`(flow: MutableStateFlow<T>, value: T) {
    flow.emit(value)
}

suspend fun <T> `~~~!!! recognizableFrameWithEmitInAsyncStackTraceAssign 4 !!!~~~`(flow: MutableStateFlow<T>, value: T) {
    flow.value = value
}

suspend fun <T> `~~~!!! recognizableFrameWithEmitInAsyncStackTraceEmit 4 !!!~~~`(flow: MutableStateFlow<T>, value: T) {
    flow.emit(value)
}
