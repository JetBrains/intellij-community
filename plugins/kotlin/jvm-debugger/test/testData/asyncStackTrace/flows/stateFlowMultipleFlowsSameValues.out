LineBreakpoint created at stateFlowMultipleFlowsSameValues.kt:20
LineBreakpoint created at stateFlowMultipleFlowsSameValues.kt:26
Run Java
Connected to the target VM
stateFlowMultipleFlowsSameValues.kt:20
vvv stack trace vvv
  emit
  <hidden frames>
-- Async stack trace --
  <hidden frames>
  ~~~!!! recognizableFrameWithEmitInAsyncStackTraceConstructor 1 !!!~~~
  <clinit>
^^^ stack trace ^^^
stateFlowMultipleFlowsSameValues.kt:26
vvv stack trace vvv
  emit
  <hidden frames>
-- Async stack trace --
  <hidden frames>
  ~~~!!! recognizableFrameWithEmitInAsyncStackTraceConstructor 2 !!!~~~
  <clinit>
^^^ stack trace ^^^
stateFlowMultipleFlowsSameValues.kt:20
vvv stack trace vvv
  emit
  <hidden frames>
-- Async stack trace --
  <hidden frames>
  ~~~!!! recognizableFrameWithEmitInAsyncStackTraceAssign 1 !!!~~~
  invokeSuspend
  <hidden frames>
  main
  main
^^^ stack trace ^^^
stateFlowMultipleFlowsSameValues.kt:26
vvv stack trace vvv
  emit
  <hidden frames>
-- Async stack trace --
  <hidden frames>
  ~~~!!! recognizableFrameWithEmitInAsyncStackTraceAssign 2 !!!~~~
  invokeSuspend
  <hidden frames>
  main
  main
^^^ stack trace ^^^
stateFlowMultipleFlowsSameValues.kt:20
vvv stack trace vvv
  emit
  <hidden frames>
-- Async stack trace --
  <hidden frames>
  ~~~!!! recognizableFrameWithEmitInAsyncStackTraceEmit 1 !!!~~~
  invokeSuspend
  <hidden frames>
  main
  main
^^^ stack trace ^^^
stateFlowMultipleFlowsSameValues.kt:26
vvv stack trace vvv
  emit
  <hidden frames>
-- Async stack trace --
  <hidden frames>
  ~~~!!! recognizableFrameWithEmitInAsyncStackTraceEmit 2 !!!~~~
  invokeSuspend
  <hidden frames>
  main
  main
^^^ stack trace ^^^
stateFlowMultipleFlowsSameValues.kt:20
vvv stack trace vvv
  emit
  <hidden frames>
-- Async stack trace --
  <hidden frames>
  ~~~!!! recognizableFrameWithEmitInAsyncStackTraceCas 1 !!!~~~
  invokeSuspend
  <hidden frames>
  main
  main
^^^ stack trace ^^^
stateFlowMultipleFlowsSameValues.kt:26
vvv stack trace vvv
  emit
  <hidden frames>
-- Async stack trace --
  <hidden frames>
  ~~~!!! recognizableFrameWithEmitInAsyncStackTraceCas 2 !!!~~~
  invokeSuspend
  <hidden frames>
  main
  main
^^^ stack trace ^^^
stateFlowMultipleFlowsSameValues.kt:20
vvv stack trace vvv
  emit
  <hidden frames>
-- Async stack trace --
  <hidden frames>
  ~~~!!! recognizableFrameWithEmitInAsyncStackTraceAssign 3 !!!~~~
  invokeSuspend
  <hidden frames>
  main
  main
^^^ stack trace ^^^
stateFlowMultipleFlowsSameValues.kt:26
vvv stack trace vvv
  emit
  <hidden frames>
-- Async stack trace --
  <hidden frames>
  ~~~!!! recognizableFrameWithEmitInAsyncStackTraceAssign 4 !!!~~~
  invokeSuspend
  <hidden frames>
  main
  main
^^^ stack trace ^^^
stateFlowMultipleFlowsSameValues.kt:20
vvv stack trace vvv
  emit
  <hidden frames>
-- Async stack trace --
  <hidden frames>
  ~~~!!! recognizableFrameWithEmitInAsyncStackTraceEmit 3 !!!~~~
  invokeSuspend
  <hidden frames>
  main
  main
^^^ stack trace ^^^
stateFlowMultipleFlowsSameValues.kt:26
vvv stack trace vvv
  emit
  <hidden frames>
-- Async stack trace --
  <hidden frames>
  ~~~!!! recognizableFrameWithEmitInAsyncStackTraceEmit 4 !!!~~~
  invokeSuspend
  <hidden frames>
  main
  main
^^^ stack trace ^^^
Disconnected from the target VM

Process finished with exit code 0
